source/extensions/stat_sinks/common/statsd/statsd.cc

```diff
@@ -38,34 +40,73 @@ void UdpStatsdSink::WriterImpl::write(const std::string& message) {
   Network::Utility::writeToSocket(*io_handle_, &slice, 1, nullptr, *parent_.server_address_);
 }
 
+void UdpStatsdSink::WriterImpl::writeBuffer(Buffer::Instance* data) {
+  Network::Utility::writeToSocket(*io_handle_, *data, nullptr, *parent_.server_address_);
+}
+
 UdpStatsdSink::UdpStatsdSink(ThreadLocal::SlotAllocator& tls,
                              Network::Address::InstanceConstSharedPtr address, const bool use_tag,
-                             const std::string& prefix)
+                             const std::string& prefix, uint64_t buffer_size)
     : tls_(tls.allocateSlot()), server_address_(std::move(address)), use_tag_(use_tag),
-      prefix_(prefix.empty() ? Statsd::getDefaultPrefix() : prefix) {
+      prefix_(prefix.empty() ? Statsd::getDefaultPrefix() : prefix), buffer_size_(buffer_size) {
   tls_->set([this](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
     return std::make_shared<WriterImpl>(*this);
   });
 }
 
 void UdpStatsdSink::flush(Stats::MetricSnapshot& snapshot) {
   Writer& writer = tls_->getTyped<Writer>();
+  auto buffer = Buffer::OwnedImpl();
+
   for (const auto& counter : snapshot.counters()) {
     if (counter.counter_.get().used()) {
-      writer.write(absl::StrCat(prefix_, ".", getName(counter.counter_.get()), ":", counter.delta_,
-                                "|c", buildTagStr(counter.counter_.get().tags())));
+      std::string counter_str =
+          absl::StrCat(prefix_, ".", getName(counter.counter_.get()), ":", counter.delta_, "|c",
+                       buildTagStr(counter.counter_.get().tags()));
+      writeBuffer(buffer, writer, counter_str);
     }
   }
 
   for (const auto& gauge : snapshot.gauges()) {
     if (gauge.get().used()) {
-      writer.write(absl::StrCat(prefix_, ".", getName(gauge.get()), ":", gauge.get().value(), "|g",
-                                buildTagStr(gauge.get().tags())));
+      std::string gauge_str =
+          absl::StrCat(prefix_, ".", getName(gauge.get()), ":", gauge.get().value(), "|g",
+                       buildTagStr(gauge.get().tags()));
+      writeBuffer(buffer, writer, gauge_str);
     }
   }
+
+  flushBuffer(buffer, writer);
   // TODO(efimki): Add support of text readouts stats.
 }
 
+void UdpStatsdSink::writeBuffer(Buffer::OwnedImpl& buffer, Writer& writer,
+                                const std::string& statsd_metric) const {
+  if (statsd_metric.length() > buffer_size_) {
```

Comment:
if (statsd_metric.length() >= buffer_size_) {

to avoid an unnecessary call to flushBuffer on an empty buffer when statsd_metric.length() == buffer_size_ on line 92

Reply:
Ahh, good catch! Thank you 😄